numalogic/connectors/druid/_druid.py

```diff
@@ -128,33 +158,130 @@ def fetch(
             aggregations=aggregations,
             post_aggregations=post_aggregations,
         )
-        try:
-            response = self.client.groupby(**query_params)
-        except Exception as err:
-            raise DruidFetcherError("Druid Exception:\n") from err
-        else:
-            df = response.export_pandas()
-            if df.empty or df.shape[0] == 0:
-                logging.warning("No data found for keys %s", filter_pairs)
-                return pd.DataFrame()
-
-            df["timestamp"] = pd.to_datetime(df["timestamp"]).astype("int64") // 10**6
-
-            if group_by:
-                df = df.groupby(by=group_by).sum().reset_index()
-
-            if pivot and pivot.columns:
-                df = df.pivot(
-                    index=pivot.index,
-                    columns=pivot.columns,
-                    values=pivot.value,
-                )
-                df.columns = df.columns.map("{0[1]}".format)
-                df.reset_index(inplace=True)
+        df = self._fetch(**query_params)
+
+        if df.empty or df.shape[0] == 0:
+            logging.warning("No data found for keys %s", filter_pairs)
+            return pd.DataFrame()
+
+        df["timestamp"] = pd.to_datetime(df["timestamp"]).astype("int64") // 10**6
+
+        if group_by:
+            df = df.groupby(by=group_by).sum().reset_index()
 
-            _end_time = time.perf_counter() - _start_time
-            _LOGGER.debug("params: %s latency: %.6fs", query_params, _end_time)
-            return df
+        if pivot and pivot.columns:
+            df = df.pivot(
+                index=pivot.index,
+                columns=pivot.columns,
+                values=pivot.value,
+            )
+            df.columns = df.columns.map("{0[1]}".format)
+            df.reset_index(inplace=True)
+
+        _end_time = time.perf_counter() - _start_time
+        _LOGGER.debug("Druid params: %s, fetch time: %.4fs", query_params, _end_time)
+        return df
 
     def raw_fetch(self, *args, **kwargs) -> pd.DataFrame:
         raise NotImplementedError
+
+    def chunked_fetch(
```

Comment:
@ab93 where is this being used?

Reply:
That will be in a different PR for the UDF change. This is just for adding the API.